File: src/database/queries.js
/**
 * Database query functions for opportunities, posts, and caching
 */

import { logger } from '../utils/logger.js';
import crypto from 'crypto';

/**
 * Generate opportunity ID from source and source_id
 * @param {string} source - Source type (e.g., 'github-crowdsource')
 * @param {string} sourceId - Source-specific ID
 * @returns {string} Safe, collision-resistant ID
 */
export function generateOpportunityId(source, sourceId) {
  // Sanitize source_id to remove any control characters that could cause collisions
  const sanitizedSourceId = sourceId.replace(/[^a-zA-Z0-9_.-]/g, '_');
  // Use ::: as delimiter (unlikely to appear in source names)
  return `${source}:::${sanitizedSourceId}`;
}

/**
 * Generate deduplication hash for an opportunity
 * @param {object} opportunity - Opportunity object
 * @returns {string} SHA-256 hash
 */
export function generateHash(opportunity) {
  const { title, company, location_text } = opportunity;
  const normalized = `${company}|${normalizeTitle(title)}|${normalizeLocation(location_text)}`;
  return crypto.createHash('sha256').update(normalized).digest('hex');
}

/**
 * Normalize title for deduplication
 * @param {string} title
 * @returns {string}
 */
function normalizeTitle(title) {
  return title
    .toLowerCase()
    .replace(/intern(ship)?|co-op|summer|fall|spring|winter|2024|2025|2026/gi, '')
    .replace(/\s+/g, ' ')
    .trim();
}

/**
 * Normalize location for deduplication
 * @param {string} location
 * @returns {string}
 */
function normalizeLocation(location) {
  return (location || '')
    .toLowerCase()
    .replace(/\s*,\s*/g, ',')
    .trim();
}

/**
 * Insert or update an opportunity
 * @param {object} db - Database client
 * @param {object} opportunity - Opportunity object
 * @returns {Promise<string>} Opportunity ID
 */
export async function upsertOpportunity(db, opportunity) {
  // Generate hash if not provided
  if (!opportunity.hash) {
    opportunity.hash = generateHash(opportunity);
  }

  // Generate ID if not provided using shared function
  const id = opportunity.id || generateOpportunityId(opportunity.source, opportunity.source_id);

  // Convert skills array to JSON string
  const skillsJson = Array.isArray(opportunity.skills)
    ? JSON.stringify(opportunity.skills)
    : opportunity.skills;

  try {
    const now = new Date().toISOString();

    // Step 1: Check for exact match by source+source_id (legitimate refresh/update)
    const exactMatch = await db.execute({
      sql: `SELECT id, first_seen, source_id FROM opportunities
            WHERE source = ? AND source_id = ?
            LIMIT 1`,
      args: [opportunity.source, opportunity.source_id],
    });

    if (exactMatch.rows.length > 0) {
      // Exact match - update existing opportunity
      const existingId = exactMatch.rows[0].id;
      const firstSeen = exactMatch.rows[0].first_seen;

      await db.execute({
        sql: `UPDATE opportunities SET
          title = ?, company = ?, type = ?, workplace_type = ?,
          location_text = ?, lat = ?, lon = ?, distance_km = ?, is_california = ?,
          url = ?, deadline = ?, posted_at = ?,
          description_raw = ?, description_preview = ?, skills = ?, compensation = ?,
          cc_friendly = ?, cc_exclusion_reason = ?,
          last_seen = ?, hash = ?, score = ?
          WHERE id = ?`,
        args: [
          opportunity.title,
          opportunity.company,
          opportunity.type,
          opportunity.workplace_type,
          opportunity.location_text,
          opportunity.lat,
          opportunity.lon,
          opportunity.distance_km,
          opportunity.is_california ? 1 : 0,
          opportunity.url,
          opportunity.deadline,
          opportunity.posted_at,
          opportunity.description_raw,
          opportunity.description_preview,
          skillsJson,
          opportunity.compensation,
          opportunity.cc_friendly ? 1 : 0,
          opportunity.cc_exclusion_reason,
          now,
          opportunity.hash,
          opportunity.score,
          existingId,
        ],
      });

      logger.debug(`Updated opportunity (exact match): ${existingId}`);
      return existingId;
    }

    // Step 2: Check for hash collision (same content, different source or source_id)
    const hashMatch = await db.execute({
      sql: `SELECT id, source, source_id FROM opportunities WHERE hash = ? LIMIT 1`,
      args: [opportunity.hash],
    });

    if (hashMatch.rows.length > 0) {
      const existingId = hashMatch.rows[0].id;
      const existingSource = hashMatch.rows[0].source;
      const existingSourceId = hashMatch.rows[0].source_id;

      // Check if same source but different source_id (GitHub data quality issue)
      if (existingSource === opportunity.source && existingSourceId !== opportunity.source_id) {
        // HASH COLLISION DETECTED - Same source, different UUID
        logger.warn('Hash collision detected - same source has duplicate UUIDs for identical content', {
          existingId,
          existingSourceId,
          newSourceId: opportunity.source_id,
          source: opportunity.source,
          title: opportunity.title,
          company: opportunity.company,
          hash: opportunity.hash,
        });

        // Return existing ID WITHOUT updating source_id (preserve first-seen UUID)
        // This prevents duplicate posts while maintaining data integrity
        return existingId;
      }

      // Different source with same hash - legitimate cross-source duplicate
      logger.info('Cross-source duplicate detected - preserving original data, updating last_seen only', {
        existingId,
        existingSource,
        newSource: opportunity.source,
        title: opportunity.title,
        company: opportunity.company,
      });

      // Update ONLY last_seen timestamp (preserve original source's data)
      // This prevents data corruption from overwriting URLs, descriptions, etc.
      await db.execute({
        sql: `UPDATE opportunities SET last_seen = ? WHERE id = ?`,
        args: [now, existingId],
      });

      logger.debug(`Updated last_seen for cross-source match: ${existingId}`);
      return existingId;
    }

    // Step 3: No match - insert new opportunity
    await db.execute({
      sql: `INSERT INTO opportunities (
        id, source, source_id, title, company, type, workplace_type,
        location_text, lat, lon, distance_km, is_california,
        url, deadline, posted_at,
        description_raw, description_preview, skills, compensation,
        cc_friendly, cc_exclusion_reason,
        first_seen, last_seen, hash, score
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      args: [
        id,
        opportunity.source,
        opportunity.source_id,
        opportunity.title,
        opportunity.company,
        opportunity.type,
        opportunity.workplace_type,
        opportunity.location_text,
        opportunity.lat,
        opportunity.lon,
        opportunity.distance_km,
        opportunity.is_california ? 1 : 0,
        opportunity.url,
        opportunity.deadline,
        opportunity.posted_at,
        opportunity.description_raw,
        opportunity.description_preview,
        skillsJson,
        opportunity.compensation,
        opportunity.cc_friendly ? 1 : 0,
        opportunity.cc_exclusion_reason,
        now,
        now,
        opportunity.hash,
        opportunity.score,
      ],
    });

    logger.debug(`Inserted new opportunity: ${id}`);
    return id;
  } catch (error) {
    logger.error('Failed to upsert opportunity', {
      opportunityId: id,
      error: error.message,
    });
    throw error;
  }
}

/**
 * Get opportunity by source and source_id
 * @param {object} db - Database client
 * @param {string} source - Source type
 * @param {string} sourceId - Source-specific ID
 * @returns {Promise<object|null>} Opportunity or null if not found
 */
export async function getOpportunityBySourceId(db, source, sourceId) {
  // Input validation
  if (!source || typeof source !== 'string') {
    throw new Error('source must be a non-empty string');
  }

  if (!sourceId || typeof sourceId !== 'string') {
    throw new Error('sourceId must be a non-empty string');
  }

  if (!db) {
    throw new Error('db client is required');
  }

  try {
    const result = await db.execute({
      sql: `SELECT * FROM opportunities WHERE source = ? AND source_id = ? LIMIT 1`,
      args: [source, sourceId],
    });

    if (result.rows.length > 0) {
      return parseOpportunityRow(result.rows[0]);
    }

    return null;
  } catch (error) {
    logger.error('Failed to get opportunity by source ID', {
      source,
      sourceId,
      error: error.message,
    });
    throw error;
  }
}

/**
 * Get existing opportunities by source and source_ids (batch query to avoid N+1)
 * @param {object} db - Database client
 * @param {string} source - Source type
 * @param {Array<string>} sourceIds - Array of source-specific IDs
 * @returns {Promise<Map<string, object>>} Map of sourceId -> opportunity
 */
export async function getExistingOpportunitiesBySourceIds(db, source, sourceIds) {
  // Input validation
  if (!source || typeof source !== 'string') {
    throw new Error('source must be a non-empty string');
  }

  if (!Array.isArray(sourceIds) || sourceIds.length === 0) {
    throw new Error('sourceIds must be a non-empty array');
  }

  if (!db) {
    throw new Error('db client is required');
  }

  // Validate all sourceIds are strings
  if (!sourceIds.every(id => typeof id === 'string' && id.length > 0)) {
    throw new Error('all sourceIds must be non-empty strings');
  }

  try {
    // Build SQL with parameterized IN clause
    const placeholders = sourceIds.map(() => '?').join(',');
    const sql = `SELECT * FROM opportunities WHERE source = ? AND source_id IN (${placeholders})`;
    const args = [source, ...sourceIds];

    const result = await db.execute({ sql, args });

    // Convert to Map for O(1) lookup
    const existingMap = new Map();
    for (const row of result.rows) {
      const parsed = parseOpportunityRow(row);
      existingMap.set(row.source_id, parsed);
    }

    logger.debug(`Batch query found ${existingMap.size} existing opportunities out of ${sourceIds.length}`);

    return existingMap;
  } catch (error) {
    logger.error('Failed to get existing opportunities by source IDs', {
      source,
      sourceIdsCount: sourceIds.length,
      error: error.message,
    });
    throw error;
  }
}

/**
 * Get unposted opportunities ordered by score
 * @param {object} db - Database client
 * @param {number} limit - Maximum number to return
 * @returns {Promise<Array>} Array of opportunities
 */
export async function getUnpostedOpportunities(db, limit = 10) {
  try {
    const result = await db.execute({
      sql: `SELECT o.*
            FROM opportunities o
            LEFT JOIN posts p ON o.id = p.opportunity_id
            WHERE p.id IS NULL
              AND o.is_california = 1
              AND o.score > 0
            ORDER BY o.score DESC
            LIMIT ?`,
      args: [limit],
    });

    return result.rows.map(parseOpportunityRow);
  } catch (error) {
    logger.error('Failed to get unposted opportunities', { error: error.message });
    throw error;
  }
}

/**
 * Mark opportunity as posted
 * @param {object} db - Database client
 * @param {string} opportunityId - Opportunity ID
 * @param {string} discordMessageId - Discord message ID
 * @param {string} channelId - Discord channel ID
 * @returns {Promise<void>}
 */
export async function markAsPosted(db, opportunityId, discordMessageId, channelId) {
  try {
    await db.execute({
      sql: `INSERT INTO posts (opportunity_id, discord_message_id, channel_id)
            VALUES (?, ?, ?)`,
      args: [opportunityId, discordMessageId, channelId],
    });

    logger.info(`Marked opportunity as posted: ${opportunityId}`);
  } catch (error) {
    // Handle UNIQUE constraint violation gracefully (duplicate post prevented by DB)
    if (error.message?.includes('UNIQUE constraint')) {
      logger.info(`Opportunity already posted (prevented by UNIQUE constraint): ${opportunityId}`, {
        channelId,
        discordMessageId,
      });
      return; // Not an error - database prevented duplicate
    }

    // Other errors are real failures
    logger.error('Failed to mark opportunity as posted', {
      opportunityId,
      error: error.message,
    });
    throw error;
  }
}

/**
 * Get opportunity statistics
 * @param {object} db - Database client
 * @returns {Promise<object>} Statistics object
 */
export async function getStats(db) {
  try {
    const [total, posted, unposted, california, lastFetch] = await Promise.all([
      db.execute('SELECT COUNT(*) as count FROM opportunities'),
      db.execute('SELECT COUNT(DISTINCT opportunity_id) as count FROM posts'),
      db.execute(`
        SELECT COUNT(*) as count FROM opportunities o
        LEFT JOIN posts p ON o.id = p.opportunity_id
        WHERE p.id IS NULL
      `),
      db.execute('SELECT COUNT(*) as count FROM opportunities WHERE is_california = 1'),
      db.execute('SELECT MAX(last_fetched) as last FROM sources'),
    ]);

    return {
      totalOpportunities: total.rows[0].count,
      posted: posted.rows[0].count,
      unposted: unposted.rows[0].count,
      california: california.rows[0].count,
      lastFetch: lastFetch.rows[0].last,
    };
  } catch (error) {
    logger.error('Failed to get stats', { error: error.message });
    throw error;
  }
}

/**
 * Find duplicate opportunities by hash
 * @param {object} db - Database client
 * @param {string} hash - Opportunity hash
 * @returns {Promise<object|null>} Existing opportunity or null
 */
export async function findDuplicateByHash(db, hash) {
  try {
    const result = await db.execute({
      sql: 'SELECT * FROM opportunities WHERE hash = ? LIMIT 1',
      args: [hash],
    });

    if (result.rows.length > 0) {
      return parseOpportunityRow(result.rows[0]);
    }

    return null;
  } catch (error) {
    logger.error('Failed to find duplicate', { error: error.message });
    throw error;
  }
}

/**
 * Update source fetch record
 * @param {object} db - Database client
 * @param {string} sourceType - Source type (greenhouse, lever, etc.)
 * @param {string} sourceKey - Source key (company identifier)
 * @param {number} successCount - Number of opportunities found
 * @param {string|null} errorMessage - Error message if failed
 * @returns {Promise<void>}
 */
export async function updateSourceFetch(db, sourceType, sourceKey, successCount, errorMessage = null) {
  try {
    const now = new Date().toISOString();

    await db.execute({
      sql: `INSERT INTO sources (source_type, source_key, last_fetched, fetch_count, last_success_count, last_error)
            VALUES (?, ?, ?, 1, ?, ?)
            ON CONFLICT(source_type, source_key) DO UPDATE SET
              last_fetched = ?,
              fetch_count = fetch_count + 1,
              last_success_count = ?,
              last_error = ?`,
      args: [
        sourceType,
        sourceKey,
        now,
        successCount,
        errorMessage,
        now,
        successCount,
        errorMessage,
      ],
    });
  } catch (error) {
    logger.error('Failed to update source fetch record', {
      sourceType,
      sourceKey,
      error: error.message,
    });
    // Don't throw - this is non-critical
  }
}

/**
 * Clean up old opportunities
 * @param {object} db - Database client
 * @param {number} retentionDays - Number of days to retain
 * @returns {Promise<number>} Number of deleted opportunities
 */
export async function cleanupOldOpportunities(db, retentionDays = 90) {
  try {
    const cutoffDate = new Date();
    cutoffDate.setDate(cutoffDate.getDate() - retentionDays);
    const cutoffIso = cutoffDate.toISOString();

    const result = await db.execute({
      sql: `DELETE FROM opportunities
            WHERE last_seen < ?
            AND id NOT IN (SELECT opportunity_id FROM posts)`,
      args: [cutoffIso],
    });

    const deleted = result.rowsAffected || 0;
    logger.info(`Cleaned up ${deleted} old opportunities (older than ${retentionDays} days)`);
    return deleted;
  } catch (error) {
    logger.error('Failed to cleanup old opportunities', { error: error.message });
    throw error;
  }
}

/**
 * Parse database row into opportunity object
 * @param {object} row - Database row
 * @returns {object} Parsed opportunity
 */
function parseOpportunityRow(row) {
  let skills = [];
  if (row.skills) {
    try {
      skills = JSON.parse(row.skills);
    } catch (error) {
      logger.warn('Failed to parse skills JSON', { skills: row.skills, error: error.message });
      skills = [];
    }
  }

  return {
    ...row,
    is_california: row.is_california === 1,
    cc_friendly: row.cc_friendly === 1,
    skills,
  };
}

File: src/database/queries.py
"""
CRUD operations for the SMC CS Opportunities Bot database.

Provides functions for inserting, querying, updating, and deleting
opportunities, posts, and cache entries.
"""

from datetime import datetime, timedelta
from typing import List, Optional, Tuple

from src.database.db import get_database
from src.database.models import GeocodeCache, Opportunity, Post, Source
from src.utils.config import get_config
from src.utils.logger import get_logger

logger = get_logger(__name__)


# ===== Opportunity Operations =====

def insert_opportunity(opp: Opportunity) -> bool:
    """
    Insert a new opportunity into the database.

    Args:
        opp: Opportunity object to insert

    Returns:
        True if inserted successfully, False if already exists
    """
    db = get_database()

    try:
        with db.get_connection() as conn:
            cursor = conn.cursor()

            opp_dict = opp.to_dict()

            cursor.execute(
                """
                INSERT INTO opportunities (
                    id, source, source_id, title, company, type, workplace_type,
                    location_text, lat, lon, distance_km, is_local, url, deadline,
                    posted_at, description_raw, summary_bullets, cc_friendly,
                    cc_reason, first_seen, last_seen, hash, score, created_at
                ) VALUES (
                    :id, :source, :source_id, :title, :company, :type, :workplace_type,
                    :location_text, :lat, :lon, :distance_km, :is_local, :url, :deadline,
                    :posted_at, :description_raw, :summary_bullets, :cc_friendly,
                    :cc_reason, :first_seen, :last_seen, :hash, :score, :created_at
                )
                """,
                opp_dict,
            )

            logger.debug(f"Inserted opportunity: {opp.title} at {opp.company}")
            return True

    except Exception as e:
        if "UNIQUE constraint failed" in str(e):
            logger.debug(f"Opportunity already exists: {opp.source}/{opp.source_id}")
            return False
        else:
            logger.error(f"Error inserting opportunity: {e}")
            raise


def update_opportunity(opp: Opportunity) -> bool:
    """
    Update an existing opportunity.

    Args:
        opp: Opportunity object with updated data

    Returns:
        True if updated successfully
    """
    db = get_database()

    with db.get_connection() as conn:
        cursor = conn.cursor()

        opp_dict = opp.to_dict()

        cursor.execute(
            """
            UPDATE opportunities SET
                title = :title,
                company = :company,
                workplace_type = :workplace_type,
                location_text = :location_text,
                lat = :lat,
                lon = :lon,
                distance_km = :distance_km,
                is_local = :is_local,
                url = :url,
                deadline = :deadline,
                posted_at = :posted_at,
                description_raw = :description_raw,
                summary_bullets = :summary_bullets,
                cc_friendly = :cc_friendly,
                cc_reason = :cc_reason,
                last_seen = :last_seen,
                score = :score
            WHERE id = :id
            """,
            opp_dict,
        )

        logger.debug(f"Updated opportunity: {opp.id}")
        return True


def get_opportunity_by_id(opp_id: str) -> Optional[Opportunity]:
    """
    Retrieve an opportunity by ID.

    Args:
        opp_id: Opportunity ID

    Returns:
        Opportunity object or None if not found
    """
    db = get_database()

    with db.get_connection() as conn:
        cursor = conn.cursor()

        cursor.execute(
            "SELECT * FROM opportunities WHERE id = ?",
            (opp_id,),
        )

        row = cursor.fetchone()

        if row:
            return Opportunity.from_row(row)
        return None


def get_opportunity_by_source(source: str, source_id: str) -> Optional[Opportunity]:
    """
    Retrieve an opportunity by source and source ID.

    Args:
        source: Source type (e.g., "greenhouse")
        source_id: ID from the source

    Returns:
        Opportunity object or None if not found
    """
    db = get_database()

    with db.get_connection() as conn:
        cursor = cursor = conn.cursor()

        cursor.execute(
            "SELECT * FROM opportunities WHERE source = ? AND source_id = ?",
            (source, source_id),
        )

        row = cursor.fetchone()

        if row:
            return Opportunity.from_row(row)
        return None


def get_unposted_opportunities(limit: Optional[int] = None) -> List[Opportunity]:
    """
    Get all opportunities that haven't been posted to Discord yet,
    ordered by score (highest first).

    Args:
        limit: Maximum number of opportunities to return

    Returns:
        List of Opportunity objects
    """
    db = get_database()

    with db.get_connection() as conn:
        cursor = conn.cursor()

        query = """
            SELECT o.*
            FROM opportunities o
            LEFT JOIN posts p ON o.id = p.opportunity_id
            WHERE p.opportunity_id IS NULL
            ORDER BY o.score DESC, o.first_seen DESC
        """

        if limit:
            query += f" LIMIT {limit}"

        cursor.execute(query)

        rows = cursor.fetchall()

        return [Opportunity.from_row(row) for row in rows]


def get_opportunities_by_hash(hash_value: str) -> List[Opportunity]:
    """
    Get all opportunities with a specific deduplication hash.

    Args:
        hash_value: Hash to search for

    Returns:
        List of matching Opportunity objects
    """
    db = get_database()

    with db.get_connection() as conn:
        cursor = conn.cursor()

        cursor.execute(
            "SELECT * FROM opportunities WHERE hash = ?",
            (hash_value,),
        )

        rows = cursor.fetchall()

        return [Opportunity.from_row(row) for row in rows]


def delete_old_opportunities(days: int = 90) -> int:
    """
    Delete opportunities older than specified days.

    Args:
        days: Delete opportunities with first_seen older than this many days

    Returns:
        Number of opportunities deleted
    """
    db = get_database()
    cutoff_date = (datetime.utcnow() - timedelta(days=days)).isoformat()

    with db.get_connection() as conn:
        cursor = conn.cursor()

        cursor.execute(
            "DELETE FROM opportunities WHERE first_seen < ?",
            (cutoff_date,),
        )

        deleted_count = cursor.rowcount
        logger.info(f"Deleted {deleted_count} old opportunities (older than {days} days)")

        return deleted_count


# ===== Post Operations =====

def insert_post(post: Post) -> int:
    """
    Record that an opportunity was posted to Discord.

    Args:
        post: Post object with opportunity_id and optional message details

    Returns:
        ID of the inserted post record
    """
    db = get_database()

    with db.get_connection() as conn:
        cursor = conn.cursor()

        post_dict = post.to_dict()

        cursor.execute(
            """
            INSERT INTO posts (
                opportunity_id, posted_at, discord_message_id, channel_id
            ) VALUES (
                :opportunity_id, :posted_at, :discord_message_id, :channel_id
            )
            """,
            post_dict,
        )

        post_id = cursor.lastrowid
        logger.debug(f"Recorded post for opportunity: {post.opportunity_id}")

        return post_id


def mark_as_posted(opportunity_id: str, discord_message_id: Optional[str] = None) -> int:
    """
    Mark an opportunity as posted (convenience function).

    Args:
        opportunity_id: ID of the opportunity
        discord_message_id: Optional Discord message ID

    Returns:
        ID of the inserted post record
    """
    post = Post(
        opportunity_id=opportunity_id,
        discord_message_id=discord_message_id,
    )

    return insert_post(post)


def is_posted(opportunity_id: str) -> bool:
    """
    Check if an opportunity has been posted to Discord.

    Args:
        opportunity_id: ID of the opportunity

    Returns:
        True if posted, False otherwise
    """
    db = get_database()

    with db.get_connection() as conn:
        cursor = conn.cursor()

        cursor.execute(
            "SELECT 1 FROM posts WHERE opportunity_id = ? LIMIT 1",
            (opportunity_id,),
        )

        return cursor.fetchone() is not None


# ===== Geocode Cache Operations =====

def get_cached_geocode(location_text: str) -> Optional[Tuple[float, float]]:
    """
    Get cached geocoding result for a location.

    Args:
        location_text: Location string (e.g., "Los Angeles, CA")

    Returns:
        Tuple of (lat, lon) if cached, None otherwise
    """
    db = get_database()

    with db.get_connection() as conn:
        cursor = conn.cursor()

        cursor.execute(
            "SELECT lat, lon FROM geocode_cache WHERE location_text = ?",
            (location_text,),
        )

        row = cursor.fetchone()

        if row:
            return (row[0], row[1])
        return None


def cache_geocode(location_text: str, lat: float, lon: float) -> bool:
    """
    Cache a geocoding result.

    Args:
        location_text: Location string
        lat: Latitude
        lon: Longitude

    Returns:
        True if cached successfully
    """
    db = get_database()

    try:
        with db.get_connection() as conn:
            cursor = conn.cursor()

            cache = GeocodeCache(
                location_text=location_text,
                lat=lat,
                lon=lon,
            )

            cache_dict = cache.to_dict()

            cursor.execute(
                """
                INSERT OR REPLACE INTO geocode_cache (
                    location_text, lat, lon, cached_at
                ) VALUES (
                    :location_text, :lat, :lon, :cached_at
                )
                """,
                cache_dict,
            )

            logger.debug(f"Cached geocode for: {location_text}")
            return True

    except Exception as e:
        logger.error(f"Error caching geocode: {e}")
        return False


# ===== Source Tracking Operations =====

def update_source_fetch(source_type: str, source_key: str, success: bool, error: Optional[str] = None):
    """
    Update source fetch history.

    Args:
        source_type: Type of source (e.g., "greenhouse")
        source_key: Source identifier (e.g., company name)
        success: Whether the fetch was successful
        error: Optional error message if fetch failed
    """
    db = get_database()

    with db.get_connection() as conn:
        cursor = conn.cursor()

        # Check if source exists
        cursor.execute(
            "SELECT fetch_count FROM sources WHERE source_type = ? AND source_key = ?",
            (source_type, source_key),
        )

        row = cursor.fetchone()

        if row:
            # Update existing source
            new_count = row[0] + 1
            cursor.execute(
                """
                UPDATE sources SET
                    last_fetched = ?,
                    fetch_count = ?,
                    last_error = ?
                WHERE source_type = ? AND source_key = ?
                """,
                (datetime.utcnow().isoformat(), new_count, error, source_type, source_key),
            )
        else:
            # Insert new source
            cursor.execute(
                """
                INSERT INTO sources (source_type, source_key, last_fetched, fetch_count, last_error)
                VALUES (?, ?, ?, 1, ?)
                """,
                (source_type, source_key, datetime.utcnow().isoformat(), error),
            )

        logger.debug(f"Updated source fetch: {source_type}/{source_key}")


# ===== Utility Functions =====

def get_all_hashes() -> set:
    """
    Get all opportunity hashes (for deduplication).

    Returns:
        Set of all hash values in database
    """
    db = get_database()

    with db.get_connection() as conn:
        cursor = conn.cursor()

        cursor.execute("SELECT DISTINCT hash FROM opportunities")

        rows = cursor.fetchall()

        return {row[0] for row in rows}


if __name__ == "__main__":
    # Test CRUD operations
    from datetime import datetime

    print("=== Testing CRUD Operations ===\n")

    # Initialize database
    from src.database.db import init_database

    init_database()

    # Create a test opportunity
    test_opp = Opportunity(
        source="test",
        source_id="test123",
        title="Test Internship",
        company="Test Company",
        type="internship",
        url="https://example.com",
        first_seen=datetime.utcnow().isoformat(),
        last_seen=datetime.utcnow().isoformat(),
        hash="test_hash_123",
    )

    # Test insert
    print("Inserting test opportunity...")
    result = insert_opportunity(test_opp)
    print(f"  Insert result: {result}")

    # Test retrieve
    print("\nRetrieving opportunity by ID...")
    retrieved = get_opportunity_by_id(test_opp.id)
    if retrieved:
        print(f"  Found: {retrieved.title} at {retrieved.company}")

    # Test unposted query
    print("\nQuerying unposted opportunities...")
    unposted = get_unposted_opportunities(limit=5)
    print(f"  Found {len(unposted)} unposted opportunities")

    # Test post tracking
    print("\nMarking as posted...")
    mark_as_posted(test_opp.id)
    print(f"  Is posted: {is_posted(test_opp.id)}")

    # Test geocode cache
    print("\nTesting geocode cache...")
    cache_geocode("Test City, CA", 34.0, -118.0)
    cached = get_cached_geocode("Test City, CA")
    print(f"  Cached coordinates: {cached}")

    print("\n✓ All CRUD operations working")
